runtime/dynlink_nat.c

```diff
@@ -71,29 +71,42 @@ CAMLprim value caml_natdynlink_open(value filename, value global)
   void *sym;
   void *dlhandle;
   char_os *p;
-
-  /* TODO: dlclose in case of error... */
+  value exn;
 
   p = caml_stat_strdup_to_os(String_val(filename));
-  caml_enter_blocking_section();
+  exn = caml_enter_blocking_section_exn();
+  if (Is_exception_result(exn)) goto cleanup1;
   dlhandle = caml_dlopen(p, 1, Int_val(global));
-  caml_leave_blocking_section();
-  caml_stat_free(p);
+  exn = caml_leave_blocking_section_exn();
```

Comment:
```suggestion
  exn = caml_leave_blocking_section_exn();
  caml_stat_free(p);
```

Why not freeing `p` here? This would make the change smaller, the lifetime of `p` shorter and the resources of the function easier to track.

Of course, you then have to free it in the exceptional case above.